common/src/main/java/com/microsoft/identity/common/internal/cache/HelloCache.kt

```diff
@@ -0,0 +1,352 @@
+//
+//  This code is licensed under the MIT License.
+//
+//  Permission is hereby granted, free of charge, to any person obtaining a copy
+//  of this software and associated documentation files(the "Software"), to deal
+//  in the Software without restriction, including without limitation the rights
+//  to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+//  copies of the Software, and to permit persons to whom the Software is
+//  furnished to do so, subject to the following conditions :
+//
+//  The above copyright notice and this permission notice shall be included in
+//  all copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+package com.microsoft.identity.common.internal.cache
+
+import android.content.Context
+import android.content.pm.PackageManager
+import android.os.Build
+import androidx.annotation.VisibleForTesting
+import com.microsoft.identity.common.java.interfaces.INameValueStorage
+import com.microsoft.identity.common.java.interfaces.IPlatformComponents
+import com.microsoft.identity.common.java.util.StringUtil
+import com.microsoft.identity.common.logging.Logger
+import java.util.concurrent.TimeUnit
+
+/**
+ * Persisted cache for the IPC hello() protocol.
+ * Use client's protocol version and the targeted app's package name and app version as a key
+ * to cache the negotiated protocol version.
+ *
+ *
+ * This means a new hello() call will ONLY be triggered only when.
+ * 1. IPC operation is invoked for the very first time.
+ * 2. Client bumps up protocol version.
+ * 3. The targeted app is updated, uninstalled, reinstalled.
+ * 4. Cache entry is expired.
+ */
+/**
+ * @param context              application context.
+ * @param protocolName         name of the protocol that invokes hello().
+ * @param targetAppPackageName package name of the app that this client will hello() with.
+ * @param components           Platform components.
+ * @param timeoutInMs         Cache entry timeout.
+ */
+open class HelloCache (
+    private val context: Context,
+    private val protocolName: String,
+    private val targetAppPackageName: String,
+    components: IPlatformComponents,
+    private val timeoutInMs: Long = sTimeoutInMs
+) {
+    companion object {
+        private val TAG = HelloCache::class.java.simpleName
+        private const val SHARED_PREFERENCE_NAME = "com.microsoft.common.ipc.hello.cache"
+        private var sIsEnabled = true
+        private val sTimeoutInMs = TimeUnit.HOURS.toMillis(4)
```

Comment:
nit: add comments on this (and maybe rename to DEFAULT_TIMEOUT_MS or something of sorts)